vsintegration/src/FSharp.Editor/Completion/FileSystemCompletion.fs

```diff
@@ -89,38 +93,61 @@ module internal FileSystemCompletion =
  
     override this.ProvideCompletionsAsync(context) =
         async {
-            let! items = FileSystemCompletion.getItems(this, context.Document, context.Position, [|".fs"; ".fsx"|], directiveRegex)
+            let defaultSearchPath = Path.GetDirectoryName context.Document.FilePath 
+            let! extraSearchPaths = 
+                if useIncludeDirectives then
+                    FileSystemCompletion.getIncludeDirectives (context.Document, context.Position)
+                else async.Return []
+            let searchPaths = defaultSearchPath :: extraSearchPaths
+            let! items = FileSystemCompletion.getItems(this, context.Document, context.Position, allowableExtensions, directiveRegex, searchPaths)
             context.AddItems(items)
         } |> CommonRoslynHelpers.StartAsyncUnitAsTask context.CancellationToken
  
-    override __.IsInsertionTrigger(text, position, _options) = FileSystemCompletion.isInsertionTrigger(text, position)
+    override __.IsInsertionTrigger(text, position, _) = FileSystemCompletion.isInsertionTrigger(text, position)
  
     override __.GetTextChangeAsync(selectedItem, ch, cancellationToken) = 
         match FileSystemCompletion.getTextChange(selectedItem, ch) with
         | Some x -> Task.FromResult(Nullable x)
         | None -> base.GetTextChangeAsync(selectedItem, ch, cancellationToken)
 
+
+type internal LoadDirectiveCompletionProvider() =
+    inherit HashDirectiveCompletionProvider("""\s*#load\s+(@?"*(?<literal>"[^"]*"?))""", [".fs"; ".fsx"], useIncludeDirectives = true)
+
 type internal ReferenceDirectiveCompletionProvider() =
-    inherit CommonCompletionProvider()
+    inherit HashDirectiveCompletionProvider("""\s*#r\s+(@?"*(?<literal>"[^"]*"?))""", [".dll"; ".exe"], useIncludeDirectives = true)
 
-    let directiveRegex = Regex("""#r\s+(@?"*(?<literal>"[^"]*"?))""", RegexOptions.Compiled ||| RegexOptions.ExplicitCapture)
- 
-    override this.ProvideCompletionsAsync(context) =
-        async {
-            let! items = FileSystemCompletion.getItems(this, context.Document, context.Position, [|".dll"; ".exe"|], directiveRegex)
-            context.AddItems(items)
-        } |> CommonRoslynHelpers.StartAsyncUnitAsTask context.CancellationToken
- 
-    override __.IsInsertionTrigger(text, position, _options) = FileSystemCompletion.isInsertionTrigger(text, position)
- 
-    override __.GetTextChangeAsync(selectedItem, ch, cancellationToken) = 
-        match FileSystemCompletion.getTextChange(selectedItem, ch) with
-        | Some x -> Task.FromResult(Nullable x)
-        | None -> base.GetTextChangeAsync(selectedItem, ch, cancellationToken)
\ No newline at end of file
+type internal IncludeDirectiveCompletionProvider() =
+    // we have to pass an extension that's not met in real life because if we pass empty list, it does not filter at all.
+    inherit HashDirectiveCompletionProvider("""\s*#I\s+(@?"*(?<literal>"[^"]*"?))""", [".impossible_extension"], useIncludeDirectives = false)
```

Comment:
@majocha I tried empty array, it does not work (no filtering at all). 

@saul I'll try \x00.